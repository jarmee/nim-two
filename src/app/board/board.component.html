<button
  type="button"
  data-test-id="execute-play-one"
  (click)="onExecutePlay(1)"
>
  1
</button>
<button
  type="button"
  data-test-id="execute-play-two"
  (click)="onExecutePlay(2)"
>
  2
</button>
<button
  type="button"
  data-test-id="execute-play-three"
  (click)="onExecutePlay(3)"
>
  3
</button>
<span>{{ amount$ | async }}</span>
<app-board-form></app-board-form>
