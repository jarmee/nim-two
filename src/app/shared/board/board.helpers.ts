import {
  Board,
  BoardDifference,
  BoardDifferences,
  Column,
  Columns,
  Path
} from "./board.model";

function createBoardDifferencesOf(
  boardA: Board,
  andBoardB: Board
): BoardDifferences {
  const sourceBoard = boardA || andBoardB;
  return Object.keys(sourceBoard)
    .map(rowKey =>
      Object.keys(sourceBoard[rowKey]).map((columnKey: string) => ({
        newColumn: findColumnOf(boardA, [rowKey, columnKey]),
        currentColumn: findColumnOf(andBoardB, [rowKey, columnKey]),
        path: [rowKey, columnKey]
      }))
    )
    .reduce(
      (flatten: [Column, Column][], element: any) => [...flatten, ...element],
      []
    );
}

function columnsWithEqualValues(difference: BoardDifference): boolean {
  return (
    !difference.newColumn ||
    !difference.currentColumn ||
    difference.newColumn.value !== difference.currentColumn.value
  );
}

export function findColumnOf(board: Board, path: Path): Column | null {
  if (!board) return null;
  if (!path || !path.length) return null;

  return path.reduce((selectedProperty: any, pathElement: string) => {
    if (!selectedProperty || !selectedProperty[pathElement]) return null;
    return selectedProperty[pathElement];
  }, board);
}

export function differenceOf(
  boardA: Board,
  andBoardB: Board
): BoardDifferences {
  if (!boardA && !andBoardB) return [];

  return createBoardDifferencesOf(boardA, andBoardB).filter(
    columnsWithEqualValues
  );
}

export const withColumnValueFalseFilter = (column: Column) => !column.value;

export function countColumnsOf(
  board: Board,
  withFilter: (column: Column) => boolean
): number {
  if (!board) return 0;

  return Object.values(board)
    .map((columns: Columns) => Object.values(columns).filter(withFilter).length)
    .reduce(
      (amount: number, columnCount: number) => (amount += columnCount),
      0
    );
}
