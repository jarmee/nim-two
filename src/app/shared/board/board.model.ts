export type Column = {
  value: boolean;
  player?: string;
};
export type Columns = Record<number, boolean>;
export type Board = {
  [key: number]: Columns;
  errorMessage: string;
};
