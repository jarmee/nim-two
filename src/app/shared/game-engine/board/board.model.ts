export enum BoardStatus {
  InProgress = 'IN_PROGRESS'
}
export type Column = {
  value: boolean;
  player?: string;
  errorMessage?: string;
};
export type Columns = Record<number, Column>;
export type Board = {
  [key: number]: Columns;
};
export type Path = string[];
export type BoardDifference = {
  newColumn: Column;
  currentColumn: Column;
  path: Path;
};
export type BoardDifferences = Array<BoardDifference>;
