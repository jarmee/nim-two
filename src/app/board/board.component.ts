import { Component } from '@angular/core';
import { Observable } from 'rxjs';
import { GameEngineFacade } from '../shared/game-engine';
import { Board } from '../shared/game-engine/board/board.model';
import { GameStatus } from '../shared/game-engine/state/state.model';
import { Player } from '../shared/game-engine/turn/turn.model';

@Component({
  selector: 'app-board',
  templateUrl: './board.component.html',
  styleUrls: ['./board.component.scss']
})
export class BoardComponent {
  amount$: Observable<number> = this.gameEngine.amount$;

  board$: Observable<Board> = this.gameEngine.board$;

  status$: Observable<GameStatus> = this.gameEngine.status$;

  selectedPlayer$: Observable<Player> = this.gameEngine.selectedPlayer$;

  messages$: Observable<string> = this.gameEngine.messages$;

  constructor(private gameEngine: GameEngineFacade) {}

  onExecutePlay(board: Board) {
    this.gameEngine.executePlay(board);
  }

  onReset() {
    this.gameEngine.reset();
  }
}
