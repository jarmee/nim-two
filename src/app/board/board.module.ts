import { NgModule } from "@angular/core";
import { CommonModule } from "@angular/common";
import { BoardComponent } from "./board.component";
import { BoardFormComponent } from "./form/board-form.component";

@NgModule({
  declarations: [BoardComponent, BoardFormComponent],
  imports: [CommonModule],
  exports: [BoardComponent]
})
export class BoardModule {}
