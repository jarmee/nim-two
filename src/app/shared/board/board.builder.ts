import { Board } from "./board.model";

export class BoardBuilder {
  board: Board = {};

  addRowWithColumns(...columns: boolean[]): BoardBuilder {
    const currentIndex = Object.keys(this.board).length++;
    this.board = {
      ...this.board,
      [currentIndex]: {
        ...columns
          .map((value: boolean, index: number) => ({
            [index]: {
              value,
              player: null
            }
          }))
          .reduce((row: any, column: any) => ({ ...row, ...column }), {})
      }
    };
    return this;
  }

  build(): Board {
    return this.board;
  }

  static create(): BoardBuilder {
    return new BoardBuilder();
  }
}
