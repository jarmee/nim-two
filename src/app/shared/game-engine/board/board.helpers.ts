import { Board, BoardDifference, BoardDifferences, Column, Columns, Path } from './board.model';

function createBoardDifferencesOf(boardA: Board, andBoardB: Board): BoardDifferences {
  const sourceBoard = boardA || andBoardB;
  return Object.keys(sourceBoard)
    .map(rowKey =>
      Object.keys(sourceBoard[rowKey]).map((columnKey: string) => ({
        newColumn: findColumnOf(boardA, [rowKey, columnKey]),
        currentColumn: findColumnOf(andBoardB, [rowKey, columnKey]),
        path: [rowKey, columnKey]
      }))
    )
    .reduce((flatten: [Column, Column][], element: any) => [...flatten, ...element], []);
}

export function areColumnsDifferentByValue(newColumn: Column, currentColumn: Column): boolean {
  if (!newColumn && !currentColumn) {
    return false;
  }
  if (!newColumn || !currentColumn) {
    return true;
  }
  return newColumn.value !== currentColumn.value;
}

export function findColumnOf(board: Board, path: Path): Column | null {
  if (!board) {
    return null;
  }
  if (!path || !path.length) {
    return null;
  }

  return path.reduce((selectedProperty: any, pathElement: string) => {
    if (!selectedProperty || !selectedProperty[pathElement]) {
      return null;
    }
    return selectedProperty[pathElement];
  }, board);
}

export function differenceOf(boardA: Board, andBoardB: Board): BoardDifferences {
  if (!boardA && !andBoardB) {
    return [];
  }

  return createBoardDifferencesOf(boardA, andBoardB).filter(({ newColumn, currentColumn }: BoardDifference) =>
    areColumnsDifferentByValue(newColumn, currentColumn)
  );
}

export const withColumnValueFalseFilter = (column: Column) => !column.value;

export function countColumnsOf(board: Board, withFilter: (column: Column) => boolean): number {
  if (!board) {
    return 0;
  }

  return Object.values(board)
    .map((columns: Columns) => Object.values(columns).filter(withFilter).length)
    .reduce((amount: number, columnCount: number) => (amount += columnCount), 0);
}

export function exchangeColumnsOf(
  board: Board,
  exchangeColumnCallback: (column: Column, path?: Path) => Column
): Board {
  if (!board) {
    return null;
  }

  return Object.keys(board)
    .map(rowKey => ({
      [rowKey]: Object.keys(board[rowKey])
        .map(columnKey => ({
          [columnKey]: exchangeColumnCallback
            ? exchangeColumnCallback(board[rowKey][columnKey], [rowKey, columnKey])
            : { ...board[rowKey][columnKey] }
        }))
        .reduce((newRow, column) => ({ ...newRow, ...column }))
    }))
    .reduce((newBoard, row) => ({ ...newBoard, ...row }), {});
}
