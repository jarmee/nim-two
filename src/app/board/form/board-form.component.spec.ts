import { async, ComponentFixture, TestBed } from "@angular/core/testing";

import { BoardFormComponent, hasChanged } from "./board-form.component";
import { BoardFormBuilderService } from "./board-form-builder.service";
import { ReactiveFormsModule } from "@angular/forms";
import { boardFactory } from "src/app/shared/board/testing/board.mock";
import { SimpleChange, SimpleChanges } from "@angular/core";
import { Board } from "src/app/shared/board/board.model";
import { Board } from "src/app/shared/board/board.model";

describe("BoardFormComponent", () => {
  const board: Board = boardFactory();
  let component: BoardFormComponent;
  let fixture: ComponentFixture<BoardFormComponent>;

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      declarations: [BoardFormComponent],
      imports: [ReactiveFormsModule],
      providers: [BoardFormBuilderService]
    }).compileComponents();
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(BoardFormComponent);
    component = fixture.componentInstance;
    component.board = board;
    fixture.detectChanges();
  });

  it("should create", () => {
    expect(component).toBeTruthy();
  });

  it("should match the snapshot", () => {
    expect(fixture).toMatchSnapshot();
  });

  describe("hasChanged", () => {
    it("should return false if both values are null", () => {
      expect(hasChanged(null, null)).toBe(false);
    });

    it("should return true if the current board is null and the previous board has a value", () => {
      expect(hasChanged(null, {})).toBe(true);
    });

    it("should return true if the current board is not null and the previous board is null", () => {
      expect(hasChanged({}, null)).toBe(true);
    });

    it("should return true if the row count of both boards is not equal", () => {
      const currentBoard = {
        0: {},
        1: {},
        2: {}
      };
      const previousBoard = {
        0: {}
      };
      expect(hasChanged(currentBoard, previousBoard)).toBe(true);
    });

    it("should return true if the row count is equal but one colum count of one row differs", () => {
      const currentBoard = {
        0: {
          0: false
        },
        1: {
          0: false,
          2: true
        },
        2: {
          0: false
        }
      };
      const previousBoard = {
        0: {
          0: false
        },
        1: {
          0: false
        },
        2: {
          0: false
        }
      };
      expect(hasChanged(currentBoard, previousBoard)).toBe(true);
    });
  });

  describe("ngOnChanges", () => {
    const changes: SimpleChanges = {
      board: new SimpleChange(null, board, true)
    };
    beforeEach(() => {
      component.board = null;
    });

    it("should set the board property", () => {
      component.ngOnChanges(changes);

      expect(component.board).toBe(board);
    });

    it("should set the form group", () => {
      component.ngOnChanges(changes);

      expect(Object.keys(component.formGroup.controls).length).toBe(
        Object.keys(board).length
      );
    });
  });
});
