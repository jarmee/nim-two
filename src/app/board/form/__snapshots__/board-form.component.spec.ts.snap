// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BoardFormComponent should match the snapshot 1`] = `
<app-board-form
  board={[Function Object]}
  formBuilder={[Function BoardFormBuilderService]}
  formGroup={[Function FormGroup]}
>
  <form
    class="ng-untouched ng-pristine ng-valid"
    ng-reflect-form="[object Object]"
    novalidate=""
  />
</app-board-form>
`;
